browser/net/brave_network_delegate_base_browsertest.cc

```diff
@@ -0,0 +1,89 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "base/path_service.h"
+#include "brave/common/brave_paths.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_commands.h"
+#include "chrome/common/chrome_paths.h"
+#include "chrome/test/base/in_process_browser_test.h"
+#include "chrome/test/base/ui_test_utils.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/web_contents_observer.h"
+#include "content/public/test/browser_test_utils.h"
+#include "net/dns/mock_host_resolver.h"
+
+using content::NavigationHandle;
+using content::WebContents;
+using content::WebContentsObserver;
+
+class RedirectObserver : public WebContentsObserver {
+ public:
+  explicit RedirectObserver(WebContents* web_contents)
+      : WebContentsObserver(web_contents) {}
+  ~RedirectObserver() override = default;
+
+  void DidFinishNavigation(NavigationHandle* handle) override {
+    const net::HttpResponseHeaders* response = handle->GetResponseHeaders();
+    if (response) {
+      bool has_sts_header = response->HasHeader("Strict-Transport-Security");
+      sts_header_for_url.insert(
+          std::pair<GURL, bool>(handle->GetURL(), has_sts_header));
+    }
+  }
+
+  bool has_sts_header(GURL url) const {
+    auto iter = sts_header_for_url.find(url);
+    if (iter != sts_header_for_url.end()) {
+      return iter->second;
+    }
+    return false;
+  }
+
+ private:
+  std::map<GURL, bool> sts_header_for_url;
+
+  DISALLOW_COPY_AND_ASSIGN(RedirectObserver);
+};
+
+class BraveNetworkDelegateBaseBrowserTest : public InProcessBrowserTest {
+ public:
+  void SetUpOnMainThread() override {
+    InProcessBrowserTest::SetUpOnMainThread();
+    host_resolver()->AddRule("*", "127.0.0.1");
+
+    brave::RegisterPathProvider();
+    base::FilePath test_data_dir;
+    base::PathService::Get(brave::DIR_TEST_DATA, &test_data_dir);
+    embedded_test_server()->ServeFilesFromDirectory(test_data_dir);
+
+    ASSERT_TRUE(embedded_test_server()->Start());
+  }
+
+  content::WebContents* active_contents() {
+    return browser()->tab_strip_model()->GetActiveWebContents();
+  }
+};
+
+IN_PROC_BROWSER_TEST_F(BraveNetworkDelegateBaseBrowserTest, FirstPartySTS) {
+  GURL third_party =
+      embedded_test_server()->GetURL("c.com", "/iframe_hsts.html");
+
+  RedirectObserver redirect_observer(active_contents());
+  ui_test_utils::NavigateToURL(browser(), third_party);
+
+  EXPECT_TRUE(redirect_observer.has_sts_header(third_party));
+}
+
+IN_PROC_BROWSER_TEST_F(BraveNetworkDelegateBaseBrowserTest, ThirdPartySTS) {
```

Comment:
I would also add a test checking that nothing is blocked if tracking protection is off